elastio-nat-provision-lambda/lambda.py

```diff
@@ -0,0 +1,348 @@
+
+    for rt_id, rt in route_tables.items():
+        for assoc in rt['Associations']:
+            if assoc['Main']:
+                main_route_table_id = rt_id
+            subnet_id = assoc.get('SubnetId')
+            if subnet_id:
+                subnet_to_route_table[subnet_id] = rt_id
+
+    for subnet_id in subnets:
+        if subnet_id not in subnet_to_route_table:
+            subnet_to_route_table[subnet_id] = main_route_table_id
+
+    public_subnets_ids = set(get_public_subnets(subnet_to_route_table, route_tables))
+    instance_route_table_id = subnet_to_route_table[instance_subnet_id]
+    instance_route_table = route_tables[instance_route_table_id]
+
+    if not public_subnets_ids:
+        print(f"No public subnets found in {instance_vpc_id}; exiting")
+        return
+
+    if instance_subnet_id in public_subnets_ids:
+        print("Instance is running in a public subnet; exiting")
+        return
+
+    all_traffic_route = get_all_traffic_route(instance_route_table)
+
+    if all_traffic_route is not None:
+        print(f"Route table already has a route for 0.0.0.0/0; exiting. {all_traffic_route}")
+        return
+
+    print("subnets:", subnets)
+    print("public_subnets_ids:", public_subnets_ids)
+    print("instance_subnet_id:", instance_subnet_id)
+
+    nat_subnet_id = choose_subnet_for_nat(subnets, public_subnets_ids, instance_subnet_id)
+
+    if nat_subnet_id is None:
+        print("Unable to find a subnet in the same availability zone; exiting")
+        return
+
+    stack_name = f"{NAT_CFN_PREFIX}{instance_vpc_id}"
+
+    if not is_stack_deployed(stack_name):
+        print(f"No existing stack found for {instance_vpc_id},"
+              f" deploying new NAT gateway stack '{stack_name}'.")
+        deploy_nat_stack(stack_name, nat_subnet_id, instance_route_table_id)
+    else:
+        print(f"Stack {stack_name} already exists or is in progress; nothing more to do.")
+
+
+def get_stack_status(stack_name):
+    try:
+        stacks = request(
+            cfn,
+            'describe_stacks',
+            'Stacks',
+            StackName=stack_name,
+        )
+        return stacks[0]['StackStatus']
+    except cfn.exceptions.ClientError as e:
+        if 'does not exist' in str(e):
+            print(f"Stack with a name {stack_name} does not exist.")
+        else:
+            print(f"Error describing stack {stack_name}: {e}")
+            print(f"Assuming the stack {stack_name} does not exist.")
+        return None
+
+
+def is_stack_deployed(stack_name):
+    stack_status = get_stack_status(stack_name)
+
+    if stack_status == 'DELETE_IN_PROGRESS':
+        print(f"The stack {stack_name} is being deleted, waiting until the deletion completes.")
+        while True:
+            time.sleep(5)
+            stack_status = get_stack_status(stack_name)
+            if stack_status != 'DELETE_IN_PROGRESS':
+                break
+
+    if stack_status is None:
+        return False
```

Comment:
Are you sure that if the stack goes from `DELETE_IN_PROGRESS` to actually deleted, the result is that the status is `None`?  Isn't there a `DELETED` status?  I could be wrong about that I just want to confirm you've tested this scenario.

Reply:
CFN does store info about deleted stacks. But it looks like `DescribeStacks` returns an error if stack is deleted:
```
aws cloudformation describe-stacks --stack-name elastio-kqaaee-exp-1716668300

An error occurred (ValidationError) when calling the DescribeStacks operation: Stack with id elastio-kqaaee-exp-1716668300 does not exist
```
I tested this on the stack reported as deleted in CFN UI in my account.

So it looks like this handling is fine because `get_stack_status` converts "does not exist" error to `None`